src/Wellcome.Dds/Utils.Aws/S3/S3Storage.cs

```diff
@@ -1,38 +1,94 @@
 ﻿using System;
+using System.IO;
+using System.Runtime.Serialization;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Threading.Tasks;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Microsoft.Extensions.Logging;
 using Utils.Storage;
 
 namespace Utils.Aws.S3
 {
     public class S3Storage : IStorage
     {
-        private string bucket;
+        private readonly ILogger<S3Storage> logger;
+        private readonly IAmazonS3 amazonS3;
 
-        public string Folder
+        public S3Storage(
+            ILogger<S3Storage> logger,
+            IAmazonS3 amazonS3
+        )
         {
-            get { return bucket; }
-            set { bucket = value; }
+            this.logger = logger;
+            this.amazonS3 = amazonS3;
         }
 
-        public void DeleteCacheFile(string fileName)
-        {
-            throw new NotImplementedException();
-        }
+        public string Container { get; set; }
 
         public ISimpleStoredFileInfo GetCachedFile(string fileName)
         {
             // This returns an object that doesn't talk to S3 unless it needs to.
             // That is, calls to LastWriteTime or Exists should be lazy.
-            throw new NotImplementedException();
+            return new S3StoredFileInfo(Container, fileName, amazonS3);
         }
+        
+        public void DeleteCacheFile(string fileName)
+        {
+            amazonS3.DeleteObjectAsync(Container, fileName);
+        }
+
 
-        public T Read<T>(ISimpleStoredFileInfo fileInfo) where T : class
+        public async Task<T> Read<T>(ISimpleStoredFileInfo fileInfo) where T : class
         {
-            throw new NotImplementedException();
+            T t = default(T);
+            try
+            {
+                var getResponse = await amazonS3.GetObjectAsync(fileInfo.Container, fileInfo.Path);
+                await using (var stream = getResponse.ResponseStream)
+                {
+                    IFormatter formatter = new BinaryFormatter();
+                    stream.Position = 0;
+                    t = formatter.Deserialize(stream) as T;
+                    stream.Close();
+                }
+                if (t == null)
+                {
+                    logger.LogError($"Attempt to deserialize '{fileInfo.Uri}' from S3 failed.");
+                }
+            }
+            catch (Exception e)
+            {
+                logger.LogError($"Attempt to deserialize '{fileInfo.Uri}' from S3 failed.", e);
```

Comment:
If logging an exception it should come before the message - [docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loggerextensions.logerror?view=dotnet-plat-ext-3.1#Microsoft_Extensions_Logging_LoggerExtensions_LogError_Microsoft_Extensions_Logging_ILogger_System_Exception_System_String_System_Object___)

You could also use structured logging rather than string interpolation here, not sure how that'll be output in kibana but if it supports it you could then feasibly search on all values for that uri.

`logger.LogError(e, "Attempt to deserialize '{Uri}' from S3 failed.", fileInfo.Uri);`